### PR TITLE
Part 2 of RFC2906 -- allow inheriting from a different `Cargo.toml`

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

[Part 1](https://github.com/rust-lang/cargo/pull/10497)

This PR focuses on inheriting from a root workspace:
- Allow inheriting from a different `Cargo.toml`
- Add in searching for a workspace root in `to_real_manifest` as needed
- Fixed problem where a package would try to pull a dependency from a workspace and specify `{ workspace = true, optional = true }` and it would not respect the `optional`
- Added tests to verify everything is in working order

Remaining implementation work for the RFC
- Correctly inherit fields that are relative paths
  - Including adding support for inheriting `license_file`,  `readme`, and path-dependencies
-  Path dependencies infer version directive
- Lock workspace dependencies and warn when unused
- Optimizations, as needed
- Evaluate any new fields for being inheritable (e.g. `rust-version`)

Problems:
- There is duplication of code that can't be removed without significant refactoring
- Potential to parse the same manifest many times when searching for a root
  - This should not happen when a `[package]` specifies its workspace
  - This should only happen if the workspace root is greater than one folder above

### CODE REVIEW

File: src/cargo/core/workspace.rs [hunk 1]
```diff
@@ -1841,3 +1845,69 @@ impl InheritableFields {
...
 }
+
+fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
+    let key = manifest_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
+    Ok(manifest)
+}
+
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
```
Comment (epage) [inline src/cargo/core/workspace.rs:1856]: Can we share some of the core logic between `Workspace::find_root` and this?

There is very specific policy common between these and no indication of the coupling

File: src/cargo/core/workspace.rs [hunk 2]
```diff
@@ -1841,3 +1845,69 @@ impl InheritableFields {
...
+        if ances_manifest_path.exists() {
+            let man = parse_manifest(&ances_manifest_path, config)?;
+            match man.workspace_config() {
+                WorkspaceConfig::Root(ref ances_root_config) => {
+                    debug!("find_root - found a root checking exclusion");
+                    if !ances_root_config.is_excluded(manifest_path) {
+                        let mut inheritable_map = config.inheritable.borrow_mut();
+                        match inheritable_map.get(&ances_manifest_path) {
+                            Some(_) => (),
+                            None => {
```
Comment (epage) [inline src/cargo/core/workspace.rs:1881]: Is there a reason this only inserts into the cache and doesn't read from it?

(feel free to ignore if we end up removing the caching from this PR)

File: src/cargo/util/toml/mod.rs [hunk 3]
```diff
@@ -2196,7 +2281,18 @@ impl<P: ResolveToPath + Clone> TomlDependency<P> {
                     label, label
                 )).map(|dep| {
                     match dep {
-                        TomlDependency::Simple(s) => TomlDependency::Simple(s),
+                        TomlDependency::Simple(s) => {
```
Comment (epage) [inline src/cargo/util/toml/mod.rs:2284]: > Fixed problem where a package would try to pull a dependency from a workspace and specify { workspace = true, optional = true } and it would not respect the optional

Personally, I'd prefer to see unrelated fixes in a separate PR or at least a separate commit to make it easier to review, `git blame`, etc.

**Review by epage — COMMENTED** [review]
Thanks for moving this forward and sorry for the delay!

File: src/cargo/core/workspace.rs [hunk 4]
```diff
@@ -1653,6 +1653,10 @@ impl WorkspaceRootConfig {
             .collect::<Result<Vec<_>, _>>()?;
         Ok(res)
     }
+
+    pub fn inheritable(&self) -> InheritableFields {
+        self.inheritable_fields.clone()
+    }
```
[the remark this answers is not in the document]
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1659]: If I remember correctly it was to due lifetime issues, we can change it to `&InheritableFields` if that seems better.

File: src/cargo/util/toml/mod.rs [hunk 5]
```diff
@@ -1283,13 +1283,40 @@ impl TomlManifest {
...
+            match workspace_config {
+                WorkspaceConfig::Root(root) => {
+                    let inheritable = root.inheritable();
+                    let mut inheritable_map = config.inheritable.borrow_mut();
+                    match inheritable_map.get(&resolved_path) {
+                        Some(_) => (),
+                        None => {
+                            inheritable_map.insert(resolved_path, inheritable.clone());
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1300]: `.entry()` completely slipped my mind. I'll change it to this.

File: src/cargo/core/workspace.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/core/workspace.rs:1659]: Yes, ideally if the caller has a lifetime issue, it should be doing the cloning.

File: src/cargo/util/config/mod.rs [hunk 6]
```diff
@@ -202,6 +202,8 @@ pub struct Config {
     /// NOTE: this should be set before `configure()`. If calling this from an integration test,
     /// consider using `ConfigBuilder::enable_nightly_features` instead.
     pub nightly_features_allowed: bool,
+
+    pub inheritable: RefCell<HashMap<PathBuf, InheritableFields>>,
```
[the remark this answers is not in the document]
Comment (Muscraft) [reply src/cargo/util/config/mod.rs:206]: If you feel that is better, I'll remove this. I only added it as it was one way to ensure we didn't parse the same manifest many times. It was suggested in my previous attempt to use a cache so I figured I would include it here as well.

[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/util/config/mod.rs:206]: In the end, I expect we will, but we should keep these focused on one thing.

File: src/cargo/core/workspace.rs (hunk 1 above)
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1856]: The biggest problem with sharing is that `Workspace::find_root` takes a `&mut self` and uses `*self.packages.load()`. One of the ways to get around this would be to delay parsing of a `Cargo.toml` into a manifest and only resolve its `workspace_config` and then fully resolve it after we have found a root workspace. This design was similar to what was done in my last attempt and [wasn't seen as a great design](https://github.com/rust-lang/cargo/pull/9684/#discussion_r728456955) but it also had an `IntermediateManifest`. If there was a way to access `.load()` or early return if it's not a workspace root this could work but it seems significantly harder and could lead to passing things around that we don't need to. 

One idea would be to make a function that only checks a `Cargo.toml`'s `workspace_config`. This would mean that you would read a manifest from the disk more than once unless you kept all partially parsed manifests somewhere to be fully done at a later time. I don't like this way but it is something to consider. 

Unless we come up with a better idea in the meantime I will look into what amount we can share between the two places and try to implement it. If you have any suggestions on how we could change this to work better I'm all ears.

File: src/cargo/util/toml/mod.rs (hunk 3 above)
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2284]: I will add it as a separate commit on here since one of the tests depends on it.

File: src/cargo/util/config/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Muscraft) [reply src/cargo/util/config/mod.rs:206]: Sounds good, I will remove the cache

File: src/cargo/core/workspace.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1659]: I will change it so that it doesn't clone there then

File: src/cargo/core/workspace.rs (hunk 1 above)
Comment (epage) [reply src/cargo/core/workspace.rs:1856]: At minimum, I'd like the linkage between the two functions clear.

We could make an iterator of possible parents
- handle `paths::ancestors(manifest_path, None).skip(2)`
- handle `target/package`
- handle `config.home() == path`

This just leaves the parsing and excluding logic to be different

Like with my other comments, I'd recommend extracting the iterator in a commit below adding the inheritable fields that then reuses that iterator.

Comment (epage) [reply src/cargo/core/workspace.rs:1856]: (note, you might not need to create a struct to `impl Iterator` but might be able to create a function that `-> impl Iterator`)

Comment (Muscraft) [reply src/cargo/core/workspace.rs:1856]: I'm not exactly sure what you mean. Does this align with what you are trying to say? 
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .filter(|path| !path.ends_with("target/package") || &config.home() != path )
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())

}
```

Comment (epage) [reply src/cargo/core/workspace.rs:1856]: Roughly.  By using `filter`, you are treating the two `break` cases as `continue`s.  `take_while` for one will fix it but I haven't thought through what would fix the other or if that will force an explicit iterator type to be needed.
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .take_while(|path| !path.ends_with("target/package"))  // now is a break
        .filter(|path| &config.home() != path )  // this is a continue, should be a break *after* yielding the element
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())
}
```

Comment (Muscraft) [reply src/cargo/core/workspace.rs:1856]: Does that fix it?
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .take_while(|path| !path.ends_with("target/package"))  // now is a break
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())
        .take_while(|pathbuf| &config.home() != &pathbuf.parent().unwrap())
}
```

Comment (epage) [reply src/cargo/core/workspace.rs:1856]: I don't think so but you can always verify it

Comment (Muscraft) [reply src/cargo/core/workspace.rs:1856]: What I put won't work, we need to be able to grab one past any of the `_while()` methods. I didn't find anything that would work for that. Do think we should make an explicit iterator type then? 

There also is a problem with filtering for 
```rust
.filter(|ances_manifest_path| ances_manifest_path.exists())
```
since you still want to break if it equals `CARGO_HOME`

Comment (epage) [reply src/cargo/core/workspace.rs:1856]: If you also didn't find anything, probably going to require an explicit iterator

Comment (Muscraft) [reply src/cargo/core/workspace.rs:1856]: Either that or some `Fn -> _` that we call in the for loop. I'm gonna go with the explicit iterator first unless you think otherwise.

File: src/cargo/core/workspace.rs [hunk 7]
```diff
@@ -1841,3 +1825,106 @@ impl InheritableFields {
...
+    debug!("find_root - pointer {}", path.display());
+    paths::normalize_path(&path)
+}
+
+fn find_root_iter<'a>(
+    manifest_path: &'a Path,
+    config: &'a Config,
+) -> impl Iterator<Item = PathBuf> + 'a {
+    LookBehind::new(paths::ancestors(manifest_path, None).skip(2))
+        .into_iter()
```
Comment (epage) [inline src/cargo/core/workspace.rs:1880]: Is `into_iter` needed?  Seems like we should already have an iterator

File: src/cargo/core/workspace.rs [hunk 8]
```diff
@@ -1841,3 +1825,106 @@ impl InheritableFields {
...
+fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
+    let key = manifest_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
+    Ok(manifest)
+}
+
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
+    find_root_iter(manifest_path, config)
+        .find_map(|ances_manifest_path| {
```
Comment (epage) [inline src/cargo/core/workspace.rs:1838]: imo an explicit loop would be easier to read.

File: src/cargo/core/workspace.rs [hunk 9]
```diff
@@ -615,43 +605,33 @@ impl<'cfg> Workspace<'cfg> {
...
-            debug!("find_root - trying {}", ances_manifest_path.display());
-            if ances_manifest_path.exists() {
-                match *self.packages.load(&ances_manifest_path)?.workspace_config() {
-                    WorkspaceConfig::Root(ref ances_root_config) => {
-                        debug!("find_root - found a root checking exclusion");
-                        if !ances_root_config.is_excluded(manifest_path) {
-                            debug!("find_root - found!");
-                            return Ok(Some(ances_manifest_path));
+        find_root_iter(manifest_path, self.config)
+            .find_map(|ances_manifest_path| {
```
Comment (epage) [inline src/cargo/core/workspace.rs:609]: imo an explicit loop would be easier to read.

File: src/cargo/util/toml/mod.rs [hunk 10]
```diff
@@ -2073,6 +2122,22 @@ impl TomlManifest {
...
+    config: &Config,
+    resolved_path: PathBuf,
+) -> CargoResult<InheritableFields> {
+    let key = resolved_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (man, _) = read_manifest(&resolved_path, source_id, config)?;
+    match man.workspace_config() {
+        WorkspaceConfig::Root(root) => Ok(root.inheritable().clone()),
+        _ => bail!(
+            "root of a workspace inferred but wasn't a root: {}",
```
Comment (epage) [inline src/cargo/util/toml/mod.rs:2135]: Just to double check, this is mostly an internal error because we'll only call this on verified inputs?

File: src/cargo/core/workspace.rs (hunk 7 above)
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1880]: It's not needed, I'll remove it.

File: src/cargo/core/workspace.rs (hunk 8 above)
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1838]: I agree, I'll change it to one

File: src/cargo/util/toml/mod.rs (hunk 10 above)
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2135]: If a `Cargo.toml` specified `workspace = ".."` but that manifest was not a workspace root, it would throw this error. It would have to take the path below when searching for a workspace root.
```rust
WorkspaceConfig::Member { root: Some(ref path_to_root) } => {
    debug!("find_root - found pointer");
    Ok(Some(read_root_pointer(&ances_manifest_path, path_to_root)))
}
```

Comment (epage) [reply src/cargo/util/toml/mod.rs:2135]: Yes, I see the test case now (should have looked for it first).  I feel like we should improve the error to make that clearer.  Here, we have several branches that get to this error, with all but one being internal.  We can either tailor this message to that one branch or we can keep make this all internal and updated `read_root_pointer` to return an error if the pointer is bad.  Putting the error in `read_root_pointer` would also allow us to tell users what manifest had the bad pointer.  Since this isn't already done, I'm unsure what repercussions this might have.

Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2135]: I pulled the error message from 
```rust
pub fn load_workspace_config(&mut self) -> CargoResult<Option<WorkspaceRootConfig>> {
    // If we didn't find a root, it must mean there is no [workspace] section, and thus no
    // metadata.
    if let Some(root_path) = &self.root_manifest {
        let root_package = self.packages.load(root_path)?;
        match root_package.workspace_config() {
            WorkspaceConfig::Root(ref root_config) => {
                return Ok(Some(root_config.clone()));
            }

            _ => bail!(
                "root of a workspace inferred but wasn't a root: {}",
                root_path.display()
            ),
        }
    }

    Ok(None)
}
```

It looks like we could just pull it out and make `read_root_pointer` check if it's a `WorkspaceConfig::Root` but that might get complicated since one uses `.load()` and the other juts parses the manifest and doesn't save it.

Comment (epage) [reply src/cargo/util/toml/mod.rs:2135]: Oh, this is existing?  Let's just leave it the same

**Review by epage — APPROVED** [review]
Thanks!
